File: bench/fastaredux/fastaredux.dart
// The Computer Language Benchmarks Game
// http://benchmarksgame.alioth.debian.org/

// contributed by Jos Hirth
// based on the Java/C# versions by Robert F. Tobler, Enotus, Isaac Gouy, and Alp Toker

import 'dart:io';
import 'dart:typed_data';
import 'dart:math' as math;

const int LINE_LENGTH = 60;
const int OUT_BUFFER_SIZE = 256 * 1024;
const int LOOKUP_SIZE = 4 * 1024;
const double LOOKUP_SCALE = (LOOKUP_SIZE - 1) / 1;
const int NEW_LINE_CHARACTER = 10;

/// Returns the code unit of the first char.
int byte (String s) => s.codeUnitAt(0);

/// Returns all code units as an Uint8List.
Uint8List bytes (String s) => new Uint8List.fromList(s.codeUnits);

class Freq {
  int c;
  double p;
  Freq(this.c, this.p);
}

class Random {
  static const int IM = 139968;
  static const int IA = 3877;
  static const int IC = 29573;
  static const double SCALE = LOOKUP_SCALE / IM;
  static int last = 42;

  static double next () {
    return SCALE * (last = (last * IA + IC) % IM);
  }
}

class Out {
  static Uint8List buf = new Uint8List(OUT_BUFFER_SIZE);
  static const int LIMIT = OUT_BUFFER_SIZE - 2 * LINE_LENGTH - 1;
  static int ct = 0;

  static void checkFlush () {
    if (ct >= LIMIT) {
      stdout.add(new Uint8List.view(buf.buffer, 0, ct));
      buf = new Uint8List(OUT_BUFFER_SIZE);
      ct = 0;
    }
  }

  static void finalFlush() {
    stdout.add(new Uint8List.view(buf.buffer, 0, ct));
  }
}

class RandomFasta {
  static final List<Freq> lookup = new List<Freq>(LOOKUP_SIZE);

  static void makeLookup (List<Freq> a) {
    for (int i = 0, j = 0; i < LOOKUP_SIZE; i++) {
      while (a[j].p < i) {
        j++;
      }
      lookup[i] = a[j];
    }
  }

  static void addLine (int bytes) {
    Out.checkFlush();
    int lct = Out.ct;
    while (lct < Out.ct + bytes) {
      double r = Random.next();
      int ai = r.toInt();
      while (lookup[ai].p < r) {
        ai++;
      }
      Out.buf[lct++] = lookup[ai].c;
    }
    Out.buf[lct++] = NEW_LINE_CHARACTER;
    Out.ct = lct;
  }

  static void make(String desc, List<Freq> a, int n) {
    makeLookup(a);

    Out.buf.setRange(Out.ct, Out.ct + desc.length, bytes(desc), 0);
    Out.ct += desc.length;

    while (n > 0) {
      int bytes = math.min(LINE_LENGTH, n);
      addLine(bytes);
      n -= bytes;
    }
  }
}

class RepeatFasta {
  static void make(String desc, Uint8List alu, int n) {
    Out.buf.setRange(Out.ct, Out.ct + desc.length, bytes(desc), 0);
    Out.ct += desc.length;

    Uint8List buf = new Uint8List(alu.length + LINE_LENGTH);
    for (int i = 0; i < buf.length; i += alu.length) {
      buf.setRange(i, i + math.min(alu.length, buf.length - i), alu, 0);
    }

    int pos = 0;
    while (n > 0) {
      int bytes = math.min(LINE_LENGTH, n);
      Out.checkFlush();

      Out.buf.setRange(Out.ct, Out.ct + bytes, buf, pos);
      Out.ct += bytes;

      Out.buf[Out.ct++] = NEW_LINE_CHARACTER;
      pos = (pos + bytes) % alu.length;
      n -= bytes;
    }
  }
}

void main(args) {
  final String ALU =
    "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGG"
    "GAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGA"
    "CCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAAT"
    "ACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCA"
    "GCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGG"
    "AGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCC"
    "AGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA";
  final List<Freq> IUB = [
    new Freq(byte('a'), 0.27),
    new Freq(byte('c'), 0.12),
    new Freq(byte('g'), 0.12),
    new Freq(byte('t'), 0.27),
    new Freq(byte('B'), 0.02),
    new Freq(byte('D'), 0.02),
    new Freq(byte('H'), 0.02),
    new Freq(byte('K'), 0.02),
    new Freq(byte('M'), 0.02),
    new Freq(byte('N'), 0.02),
    new Freq(byte('R'), 0.02),
    new Freq(byte('S'), 0.02),
    new Freq(byte('V'), 0.02),
    new Freq(byte('W'), 0.02),
    new Freq(byte('Y'), 0.02)
  ];
  final List<Freq> HomoSapiens = [
    new Freq(byte('a'), 0.3029549426680),
    new Freq(byte('c'), 0.1979883004921),
    new Freq(byte('g'), 0.1975473066391),
    new Freq(byte('t'), 0.3015094502008)
  ];

  void sumAndScale(List<Freq> a) {
    double p = 0.0;
    for (int i = 0; i < a.length; i++) {
      a[i].p = (p += a[i].p) * LOOKUP_SCALE;
    }
    a[a.length - 1].p = LOOKUP_SCALE;
  }

  int n = args.length > 0 ? int.parse(args[0]) : 200;

  sumAndScale(IUB);
  sumAndScale(HomoSapiens);

  RepeatFasta.make(">ONE Homo sapiens alu\n", bytes(ALU), n * 2);
  RandomFasta.make(">TWO IUB ambiguity codes\n", IUB, n * 3);
  RandomFasta.make(">THREE Homo sapiens frequency\n", HomoSapiens, n * 5);
  Out.finalFlush();
}

File: bench/fastaredux/fastaredux.swift
/* The Computer Language Benchmarks Game
 http://benchmarksgame.alioth.debian.org/
 contributed by Ralph Ganszky
 converted to Swift 3 by Sergo Beruashvili
 
 Swift port of Go #3 implementation
 */

import Glibc
//import Darwin

struct AminoAcid {
    var sym: UInt8
    var prob: Double
    var cprob: Double
    
    init() {
        sym = 0
        prob = 0
        cprob = 0
    }
    
    init(_ probability: Double, _ symbol: String) {
        prob = probability
        sym = UInt8(symbol.utf8CString[0])
        cprob = 0
    }
}

let IM = 139968
let IA = 3877
let IC = 29573
var seed = 42

let n: Int
if CommandLine.arguments.count > 1 {
    n = Int(CommandLine.arguments[1]) ?? 1000
} else {
    n = 1000
}

let bufferSize = 256*1024
let width = 60
let lookupSize = 4096
let lookupScale: Double = Double(lookupSize - 1)

let aluString = "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGG" +
    "GAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGA" +
    "CCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAAT" +
    "ACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCA" +
    "GCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGG" +
    "AGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCC" +
"AGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA"
var alu = aluString.utf8CString.map({ UInt8($0) })
alu.popLast()	// Remove trailing 0-byte

var iub = [
    AminoAcid(0.27, "a"),
    AminoAcid(0.12, "c"),
    AminoAcid(0.12, "g"),
    AminoAcid(0.27, "t"),
    AminoAcid(0.02, "B"),
    AminoAcid(0.02, "D"),
    AminoAcid(0.02, "H"),
    AminoAcid(0.02, "K"),
    AminoAcid(0.02, "M"),
    AminoAcid(0.02, "N"),
    AminoAcid(0.02, "R"),
    AminoAcid(0.02, "S"),
    AminoAcid(0.02, "V"),
    AminoAcid(0.02, "W"),
    AminoAcid(0.02, "Y"),
]

var homosapiens = [
    AminoAcid(0.3029549426680, "a"),
    AminoAcid(0.1979883004921, "c"),
    AminoAcid(0.1975473066391, "g"),
    AminoAcid(0.3015094502008, "t"),
]

func computeLookup( acid: inout [AminoAcid]) -> [Int] {
    var lookup = [Int](repeating: 0, count: lookupSize)
    var p = 0.0
    for (i, _) in acid.enumerated() {
        p += acid[i].prob
        acid[i].cprob = p * lookupScale
    }
    acid[acid.count-1].cprob = 1.0 * lookupScale
    
    var j = 0
    for (i, _) in lookup.enumerated() {
        while acid[j].cprob < Double(i) {
            j += 1
        }
        lookup[i] = j
    }
    return lookup
}

func writeRBuffer(buffer: UnsafePointer<UInt8>, len: Int) {
    writeBuffer(buffer: UnsafeMutablePointer<UInt8>(mutating: buffer), len: len)
}

func writeBuffer(buffer: UnsafeMutablePointer<UInt8>,len: Int) {
    let iov = UnsafeMutablePointer<iovec>.allocate(capacity: 1)
    defer {
        iov.deallocate(capacity: 1)
    }
    iov[0].iov_base = unsafeBitCast(buffer, to: UnsafeMutableRawPointer.self)
    iov[0].iov_len = len
    writev(STDOUT_FILENO, iov, 1)
}

func randomFasta( acid: inout [AminoAcid], n: Int) {
    var cnt = n
    let lookup = computeLookup(acid: &acid)
    let buffer = UnsafeMutablePointer<UInt8>.allocate(capacity: bufferSize)
    defer {
        buffer.deallocate(capacity: bufferSize)
    }
    var pos = 0
    while cnt > 0 {
        var m = cnt
        if m > width {
            m = width
        }
        let f = lookupScale / Double(IM)
        var myrand = seed
        for _ in 0..<m {
            myrand = (myrand * IA + IC) % IM
            let r = Double(myrand) * f
            var a = lookup[Int(r)]
            while acid[a].cprob < r {
                a += 1
            }
            buffer[pos] = acid[a].sym
            pos += 1
            if pos == bufferSize {
                writeBuffer(buffer: buffer, len: pos)
                pos = 0
            }
        }
        seed = myrand
        buffer[pos] = 10
        pos += 1
        if pos == bufferSize {
            writeBuffer(buffer: buffer, len: pos)
            pos = 0
        }
        cnt -= m
    }
    if pos > 0 {
        writeBuffer(buffer: buffer, len: pos)
    }
}

func repeatFasta(gene: [UInt8], n: Int) {
    let gene2 = gene + gene
    let buffer = UnsafeMutablePointer<UInt8>.allocate(capacity: bufferSize)
    defer {
        buffer.deallocate(capacity: bufferSize)
    }
    var pos = 0
    var rpos = 0
    var cnt = n
    var lwidth = width
    while cnt > 0 {
        if pos + lwidth > bufferSize {
            writeBuffer(buffer: buffer, len: pos)
            pos = 0
        }
        if rpos + lwidth > gene.count {
            rpos = rpos % gene.count
        }
        if cnt < lwidth {
            lwidth = cnt
        }
        for i in 0..<lwidth {
            buffer[pos+i] = gene2[rpos+i]
        }
        buffer[pos+lwidth] = 10
        pos += lwidth + 1
        rpos += lwidth
        cnt -= lwidth
    }
    if pos > 0 && pos < bufferSize {
        buffer[pos] = 10
        writeBuffer(buffer: buffer, len: pos)
    } else if pos == bufferSize {
        writeBuffer(buffer: buffer, len: pos)
        buffer[0] = 10
        writeBuffer(buffer: buffer, len: 1)
    }
}

let one = ">ONE Homo sapiens alu\n"

writeRBuffer(buffer: one.utf8CString.map({ UInt8($0) }), len: one.utf8CString.count - 1)
repeatFasta(gene: alu, n: 2*n)

let two = ">TWO IUB ambiguity codes\n"
writeRBuffer(buffer: two.utf8CString.map({ UInt8($0) }), len: two.utf8CString.count - 1)
randomFasta(acid: &iub, n: 3*n)

let three = ">THREE Homo sapiens frequency\n"
writeRBuffer(buffer: three.utf8CString.map({ UInt8($0) }), len: three.utf8CString.count - 1)
randomFasta(acid: &homosapiens, n: 5*n)

